### DISCUSSION TITLE
Logging missed hits

### DISCUSSION THREAD

**Discussion by drekka** [hummingbird-project/hummingbird#130] [Q&A]
So I've setup a number of responses in Hummingbird. Now I'm trying to log the times when a request comes in and no suitable response is found. How can I do that? I tried adding a logging middleware but it only seems to be called if there is a responder found.

**Comment by adam-fowler** [comment]
Can you post the code for your middleware and also the code for when you add it to your application

**Comment by drekka** [comment]
Currently I'm doing this:

```swift
struct RequestLogger: HBMiddleware {
    func apply(to request: HBRequest, next: HBResponder) -> EventLoopFuture<HBResponse> {
        os_log(.debug, log: .mocking, "👻 %@: Received request: %@ %@", String(describing: type(of: self)), String(describing: request.method), request.uri.path)
        return next.respond(to: request)
    }
}

// Then after starting the server
server.middleware.add(RequestLogger())
```

**Reply by drekka** [reply]
It's working when the server matches, but not when it doesn't.

**Reply by adam-fowler** [reply]
What other middleware do you have attached to the application? Have you set the logLevel of HBApplication.logger to display `debug` level logging eg `HBApplication.logger.logLevel = .debug`

**Reply by adam-fowler** [reply]
Also there is a route logger middleware already available in Hummingbird: `HBLogRequestsMiddleware`

**Comment by drekka** [comment]
That's the only middleware I'm using. Looking at using the debugging level. Originally added this middleware only to have the logging look the same as the rest of the code. So I'll take another look at `HBLogRequestMiddleware` as well.

**Reply by adam-fowler** [reply]
Here is a small example of a app where the logging gets called for every request regardless of whether there is a valid route. Replacing `HBLogRequestsMiddleware` should work fine here as well

```swift
@main
struct App {
    static func main() throws {
        let app = HBApplication()
        app.logger.logLevel = .debug
        app.middleware.add(HBLogRequestsMiddleware(.debug))
        app.router.get("/health") { _ -> HTTPResponseStatus in
            return .ok
        }
        try app.start()
        app.wait()
    }
}
```

**Comment by drekka** [comment]
OK, I think I've just worked this out. I was adding my middleware after starting the server, assuming that it didn't make any difference. But as I've just found out, if I do that then the responder chain inside the server is setup to call the `trieRouter` first before calling the middleware. If I add the middleware before starting the server then it is called before the `trieRouter`. 

This appears to be related to the setup done in `HBMiddlewareGroup.constructResponder(finalResponder:)` function.

I had another look at the doco and it doesn't say anything about this (That I could find) so can I suggest adding to the middleware doco to outline that for it to be called before the router it must be added before starting the server.

**Reply by adam-fowler** [reply]
Yes your middleware would not register if it was added after the server was started. I'll add something to the docs. Editing the application in anyway once it has started will produce you defined results.
